python/activator/middleware_interface.py

```diff
@@ -1190,26 +1190,20 @@ def _filter_calibs_by_date(butler: Butler,
     filtered_calibs : iterable [`lsst.daf.butler.DatasetRef`]
         The subset of ``unfiltered_calibs`` that is valid on ``date``.
     """
-    dataset_types = {ref.datasetType for ref in unfiltered_calibs}
-    associations = {}
-    for dataset_type in dataset_types:
-        associations.update(
-            (a.ref, a) for a in butler.registry.queryDatasetAssociations(
-                dataset_type, collections, collectionTypes={CollectionType.CALIBRATION}, flattenChains=True
-            )
-        )
-
-    t = astropy.time.Time(date, scale='utc')
+    # Unfiltered_calibs can have up to one copy of each calib per certify cycle.
+    # Minimize redundant queries to find_dataset.
+    unique_ids = {(ref.datasetType, ref.dataId) for ref in unfiltered_calibs}
+    t = Timespan.fromInstant(astropy.time.Time(date, scale='utc'))
     _log_trace.debug("Looking up calibs for %s in %s.", t, collections)
-    # DatasetAssociation.timespan guaranteed not None
     filtered_calibs = []
-    for ref in unfiltered_calibs:
-        if ref in associations:
-            if associations[ref].timespan.contains(t):
-                filtered_calibs.append(ref)
-                _log_trace.debug("%s (valid over %s) matches %s.", ref, associations[ref].timespan, t)
-            else:
-                _log_trace.debug("%s (valid over %s) does not match %s.", ref, associations[ref].timespan, t)
-        else:
-            _log_trace.debug("No calib associations for %s.", ref)
+    for dataset_type, data_id in unique_ids:
+        # Use find_dataset to simultaneously filter by validity and chain order
+        found_ref = butler.find_dataset(dataset_type,
+                                        data_id,
+                                        collections=collections,
+                                        timespan=t,
+                                        dimension_records=True,
```

Comment:
I can't tell without more context, but it may be worth seeing if you can get away with setting this to `False`.  Calibrations shouldn't need any dimensions beyond what you get from registering an instrument.

Reply:
No, we need expanded refs for `transfer_from`. Worse, the requirement only causes problems in production (because it's triggered by ObsCore).